flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CoordinatorEventsToStreamOperatorRecipientExactlyOnceITCase.java

```diff
@@ -0,0 +1,385 @@
+            extends EventReceivingOperatorFactory<IN, OUT> {
+        public FailingCheckpointOperatorFactory(String name, int numEvents, int delay) {
+            super(name, numEvents, delay);
+        }
+
+        @Override
+        public <T extends StreamOperator<OUT>> T createStreamOperator(
+                StreamOperatorParameters<OUT> parameters) {
+            EventReceivingOperator<OUT> operator = new FailingCheckpointOperator<>(numEvents);
+            operator.setup(
+                    parameters.getContainingTask(),
+                    parameters.getStreamConfig(),
+                    parameters.getOutput());
+            parameters
+                    .getOperatorEventDispatcher()
+                    .registerEventHandler(parameters.getStreamConfig().getOperatorID(), operator);
+            return (T) operator;
+        }
+    }
+
+    /**
+     * A subclass of {@link EventReceivingOperator} whose subtask would fail during a specific
+     * checkpoint.
+     */
+    private static class FailingCheckpointOperator<T> extends EventReceivingOperator<T> {
+
+        private final int failAtCheckpointAfterMessage;
+
+        private TestScript testScript;
+
+        private FailingCheckpointOperator(int numEvents) {
+            this.failAtCheckpointAfterMessage =
+                    numEvents * 2 / 3 + new Random().nextInt(numEvents / 6);
+        }
+
+        @Override
+        public void setup(
+                StreamTask<?, ?> containingTask,
+                StreamConfig config,
+                Output<StreamRecord<T>> output) {
+            super.setup(containingTask, config, output);
+            if (containingTask.getIndexInSubtaskGroup() == 0) {
+                this.testScript = TestScript.getForOperator(getOperatorID() + "-subtask0");
+            } else {
+                this.testScript = null;
+            }
+        }
+
+        @Override
+        public void handleOperatorEvent(OperatorEvent evt) {
+            if (evt instanceof IntegerEvent) {
+                counter.add(1L);
+            } else if (evt instanceof EndEvent) {
+                try {
+                    state.update(Collections.singletonList(counter.getLocalValue()));
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+                if (testScript.hasAlreadyFailed()) {
+                    shouldCloseSource.set(true);
+                }
+            } else {
+                throw new UnsupportedOperationException();
+            }
+        }
+
+        @Override
+        public void snapshotState(StateSnapshotContext context) throws Exception {
+            super.snapshotState(context);
+            if (counter.getLocalValue() > failAtCheckpointAfterMessage
+                    && testScript != null
+                    && !testScript.hasAlreadyFailed()) {
+                testScript.recordHasFailed();
+                throw new RuntimeException();
```

Comment:
Have you verified that each invocation of `testFailingCheckpoint()` always triggers this exception before this test ends? Note that we typically don't want to have flaky tests that fail with low probability.

Reply:
It is guaranteed that the exception will be thrown before the test job ends, or `shouldCloseSource` would not be set to true. I'll also add an `assertThat` command to verify this point.